phabricatoremails/mail.py

```diff
@@ -0,0 +1,153 @@
+
+
+class FsMail:
+    """Writes emails to the file system
+
+    Outputs the HTML body, the text body, and the entire MIME message (as an EML file).
+    Is most useful for local development for iterating on rendering templates.
+    """
+
+    _logger: Logger
+    _from_address: str
+    _index: int
+    _output_path: pathlib.Path
+    _eml_path: pathlib.Path
+    _html_path: pathlib.Path
+    _text_path: pathlib.Path
+
+    def __init__(self, from_address: str, logger: Logger, output_path: pathlib.Path):
+        self._logger = logger
+        self._from_address = from_address
+        self._index = 0
+
+        self._output_path = output_path
+        self._eml_path = output_path / "eml"
+        self._html_path = output_path / "html"
+        self._text_path = output_path / "text"
+        self._eml_path.mkdir(parents=True, exist_ok=True)
+        self._html_path.mkdir(parents=True, exist_ok=True)
+        self._text_path.mkdir(parents=True, exist_ok=True)
+
+    def send(self, emails: List[OutgoingEmail]):
+        for email in emails:
+            basefilename = f"{self._index}-to-{email.to}"
+            with (self._eml_path / (basefilename + ".eml")).open("w") as file:
+                file.write(email.to_mime_message(self._from_address).as_string())
+            with (self._html_path / (basefilename + ".html")).open("w") as file:
+                file.write(email.html_contents)
+            with (self._text_path / (basefilename + ".text")).open("w") as file:
+                file.write(email.text_contents)
+
+            self._index += 1
+        self._logger.debug(
+            f'Recorded {len(emails)} emails to the "{self._output_path}" directory'
+        )
+
+
+@dataclass
+class SmtpMail:
+    """Sends emails via SMTP
+
+    This is useful for end-to-end testing emails in real email clients when an
+    SMTP server is available for use, but an Amazon SES account is not.
+    """
+
+    _server: smtplib.SMTP
+    _from_address: str
+    _logger: Logger
+    _send_to: Optional[str]
+
+    def send(self, emails: List[OutgoingEmail]):
+        for email in emails:
+            self._logger.debug(
+                f'[{email.to}] Sending "{email.template_name}" for "{email.subject}"'
+            )
+
+            mime_message = email.to_mime_message(
+                self._from_address, include_target_in_subject=self._send_to is not None
+            )
+            self._server.sendmail(
+                self._from_address,
+                self._send_to if self._send_to else email.to,
+                mime_message.as_string(),
+            )
+
+
+@dataclass
+class SesMail:
+    """Sends emails via Amazon SES"""
+
+    _client: Any
+    _from_address: str
+    _logger: Logger
+    _send_to: Optional[str]
+
+    @staticmethod
+    def from_aws_credentials(
+        from_address: str,
+        logger: Logger,
+        send_to: Optional[str],
+        aws_access_key_id: Optional[str],
+        aws_secret_access_key: Optional[str],
```

Comment:
`boto3.client` gathers parameters from several sources: parameters, environment variables, checking for `~/.aws/`, and perhaps more.

The `boto3.client` library encourages implicit flexibility here, where it will broaden its search if the parameters are provided, but set to `None`. Perhaps it even gracefully handles when a subset of parameters are provided.

Since the `boto3.client` API explicitly handles this "optional-y" case, I **don't** think that I should explicitly
```
if key_id or access_key:
    boto3.client("ses", key_id=key_id, access_key=access_key)
else:
    # Use environment
    boto3.client("ses")
```

Though the code itself is ambiguous, that's why there's an explicit docstring stating:
> `Uses aws credentials either from parameters, or from the environment.`

Reply:
>The first option for providing credentials to boto3 is passing them as parameters when creating clients or when creating a Session. For example:

> ```
> import boto3
> client = boto3.client(
>     's3',
>     aws_access_key_id=ACCESS_KEY,
>     aws_secret_access_key=SECRET_KEY,
>     aws_session_token=SESSION_TOKEN,
> )
> 
> # Or via the Session
> session = boto3.Session(
>     aws_access_key_id=ACCESS_KEY,
>     aws_secret_access_key=SECRET_KEY,
>     aws_session_token=SESSION_TOKEN,
> )
> ```

> where ACCESS_KEY, SECRET_KEY and SESSION_TOKEN are variables that contain your access key, secret key, and optional session token.

This does not specify that the variables can be `None`, in fact it's pretty clear that they expect those variables to contain your credentials.

> aws_access_key_id (string) -- The access key to use when creating the client. This is entirely optional, and **if not provided**, the credentials configured for the session will automatically be used. You only need to provide this argument if you want to override the credentials used for this specific client.

Reply:
[The documentation for `boto.client(...)`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/boto3.html#boto3.client) refers to [the documentation for `Session.client(...)`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.client) which shows that the `aws_` parameters are defaulted to `None`.

Reply:
Ah gotcha. I get what you're saying now, since boto isn't looking at extra keyword arguments when handling this, it will work as expected -- though interpreting the documentation, it tells the user to provide real values if they are using these keyword arguments, so it remains a little iffy. I think I get why I was thrown off -- the method signature doesn't provide `None` as the default (lines 136 and 137), and thus this pass-thru logic is a little corky. Should it be the below instead?

```suggestion
        aws_access_key_id: Optional[str] = None,
        aws_secret_access_key: Optional[str] = None,
```

Otherwise makes sense to me.

Reply:
Sure, sounds good

Reply:
(based on a previous offline conversation) This is a good example of where type hinting "first" can be tricky -- 

```python
        aws_access_key_id: Optional[str],
        aws_secret_access_key: Optional[str],
```

The original code above probably passed your IDE's checks, even though those variables are actually not optional (and a user would have to pass `None` explicitly in order to trigger the boto client to check the environment variables (or other sources) for the credentials.

The updated code looks good, though the original `Optional` hints were appropriate if you want to keep them (as per the original suggestion) for consistent hinting.

Reply:
Oh, I was following a PEP's advise that if you have `arg: Type = None`, then it knows that it's actually `arg: Optional[Type] = None` under the hood.

But, looking further into it, it looks like PEP 484 [updates the preferred convention](https://www.python.org/dev/peps/pep-0484/#union-types)

> A past version of this PEP allowed type checkers to assume an optional type when the default value is None, as in this code:

Sounds good!